### PR TITLE
Allow logger override

### PR DESCRIPTION
**Pull request by cameron-p-m**
## Problem

Ideally libraries don't have a logger at all. But this is a smaller change by allowing a user to set the Logger. This will allow the previous behaviour if needed.

Closes: https://github.com/go-mysql-org/go-mysql/issues/698 and maybe others

## Feedback?

Should errors ever log or just return?

### CODE REVIEW

File: canal/sync.go [hunk 1]
```diff
@@ -107,7 +105,6 @@ func (c *Canal) runSyncBinlog() error {
 				if e != ErrExcludedTable &&
 					e != schema.ErrTableNotExist &&
 					e != schema.ErrMissingTableMeta {
-					log.Errorf("handle rows event at (%s, %d) error %v", pos.Name, curPos, err)
```
Comment (skoef) [inline canal/sync.go:110]: why not replace this with a call to `c.cfg.Logger.Errorf`?

Comment (cameron-p-m) [reply canal/sync.go:110]: I can if you want!

Comment (cameron-p-m) [reply canal/sync.go:110]: I don't think we should log errors in the library, just return them

Comment (skoef) [reply canal/sync.go:110]: that seems like the consistent thing to do

Comment (cameron-p-m) [reply canal/sync.go:110]: sure I will change

**Comment by lance6716** [comment]
thanks for your pr! I'm quite busy today, I will take a look tomorrow.

File: canal/config.go [hunk 2]
```diff
@@ -124,5 +128,7 @@ func NewDefaultConfig() *Config {
 	c.Dump.DiscardErr = true
 	c.Dump.SkipMasterData = false
 
+	c.Logger = log.NewDefault(&log.NullHandler{})
```
Comment (lance6716) [inline canal/config.go:131]: Can you use the handler from `NewStreamHandler(os.Stdout)` (or also file a PR to siddontang/golog to make newStdHandler public)? so we won't break compatibility and users like you can still change the logger.

**Review by lance6716 — COMMENTED** [review]
I'll remove "close #632" in PR description, because other place such as `binlogsyncer.go` still access the default logger.

**Comment by cameron-p-m** [comment]
@lance6716 I made the same change in binlog syncer too now

**Review by lance6716 — COMMENTED** [review]
found that there's still some reference to global logger, like go-mysqldump, https://github.com/cameron-p-m/go-mysql/blob/fix/allow-setting-logger/mysql/mariadb_gtid.go#L96 . But it's OK to left them for a future PR.

unit tests failed, rest lgtm

@skoef do you have any comment?

**Comment by skoef** [comment]
I'm fine with this!

Perhaps sometimes we can refactor using `siddontang/go-log` and have it replaced with something more generic/mainstream as it seems that `siddontang/go-log` is no longer maintained. But that's beyond the scope of this PR!

**Comment by cameron-p-m** [comment]
@lance6716 updated and fixed tests

**Review by lance6716 — APPROVED** [review]

**Comment by naughtyGitCat** [comment]
@lance6716 when we can use this feature, v1.6? is there any way to use latest master code

**Comment by lance6716** [comment]
@atercattus Do you have a release plan?

@naughtyGitCat Currently you can specify the master branch or commit hash in go.mod https://go.dev/ref/mod#vcs-branch

**Comment by atercattus** [comment]
@lance6716, I'll do it soon.

**Comment by atercattus** [comment]
Done https://github.com/go-mysql-org/go-mysql/releases/tag/v1.6.0

**Pull request merged**